osquery/tables/system/windows/registry.cpp

```diff
@@ -266,33 +267,148 @@ void queryKey(const std::string& keyPath, QueryData& results) {
-  if ((context.constraints["key"].exists(EQUALS) &&
-       context.constraints["key"].getAll(EQUALS).size() > 0)) {
-    rKeys = context.constraints["key"].getAll(EQUALS);
-    shouldWarnLocalUsers = true;
+inline void populateDefaultKeys(std::set<std::string>& rKeys) {
+  boost::copy(kRegistryHives | boost::adaptors::map_keys,
+              std::inserter(rKeys, rKeys.end()));
+}
+
+inline void replaceKeysWithSubkeys(std::set<std::string>& rKeys) {
+  std::set<std::string> newKeys;
+  for (const auto& key : rKeys) {
+    QueryData regResults;
+    queryKey(key, regResults);
+    for (const auto& r : regResults) {
+      if (r.at("type") == "subkey") {
+        newKeys.insert(r.at("path"));
+      }
+    }
+  }
+  rKeys = newKeys;
+}
+
+inline void appendSubkeyToKeys(const std::string& subkey,
+                               std::set<std::string>& rKeys) {
+  std::set<std::string> newKeys{};
+  for (auto& key : rKeys) {
+    newKeys.insert(key + kRegSep + subkey);
+  }
+  rKeys = newKeys;
+}
+
+inline Status populateAllKeysRecursive(
+    std::set<std::string>& rKeys,
+    size_t currDepth = 1,
+    size_t maxDepth = kRegMaxRecursiveDepth) {
+  std::set<std::string> subkeys{};
+
+  if (currDepth > maxDepth) {
+    return Status(1, "Max recursive depth reached");
+  }
+
+  for (const auto& key : rKeys) {
+    QueryData regResults;
+    queryKey(key, regResults);
+    for (const auto& r : regResults) {
+      if (r.at("type") == "subkey") {
+        subkeys.insert(r.at("path"));
+      }
+    }
+  }
+
+  if (!(subkeys.size() == 0)) {
+    auto status = populateAllKeysRecursive(subkeys, ++currDepth);
+    if (!status.ok()) {
+      return status;
+    }
+    rKeys.insert(subkeys.begin(), subkeys.end());
+  }
+
+  return Status(0, "OK");
+}
+
+Status expandRegistryGlobs(const std::string& pattern,
+                           std::set<std::string>& results) {
+  auto pathElems = osquery::split(pattern, kRegSep);
+  if (pathElems.size() == 0) {
+    return Status(0, "OK");
+  }
+
+  // Pattern is '%%', grab everything
+  // Note that if '%%' is present but not at the end of the pattern
+  // , then it is treated like a single glob
+  if (boost::ends_with(pathElems[0], kSQLGlobRecursive) &&
+      pathElems.size() == 1) {
+    populateDefaultKeys(results);
+    return populateAllKeysRecursive(results);
+  }
+
+  // Special handling to insert default keys when glob present in first elem
+  if (pathElems[0].find(kSQLGlobWildcard) != std::string::npos) {
+    populateDefaultKeys(results);
+    pathElems.erase(pathElems.begin());
   } else {
-    for (auto& h : kRegistryHives) {
-      rKeys.insert(h.first);
+    results.insert(pathElems[0]);
+    pathElems.erase(pathElems.begin());
+  }
+
+  for (const auto& elem : pathElems) {
+    // We only care about  a recursive glob if it comes at the end of the
+    // pattern i.e. 'HKEY_LOCAL_MACHINE\SOFTWARE\%%'
+    if (boost::ends_with(elem, kSQLGlobRecursive) &&
+        &elem == &pathElems.back()) {
```

Comment:
I want to check if this element is the last one in the set. Is there a better way to do this? Aside from, like, doing a for loop and checking the index against pathElems.Size()?